### PR TITLE
[ENG-2378] Fix moderator permissions

### PR DESCRIPTION
**Pull request by jwalz**
## Purpose
1) Ensure that Provider Admins are also treated as Moderators
2) Update Registration permissions to give Moderators read access once the registration is pending

## Changes

To accomplish (1), I replaced the generic checks to see if the user was in the 'moderator' group with permissions checks on the provider (which are granted to both Moderators and Admins)

To accomplish (2), I extended the `can_view` function on the registration to check if the Registration is in a moderator-visible state and, if so, if the user has `view_submissions` permissions on the provider.

Wrote tests to confirm that the expected API paths are now `get`-able from moderator-visible states but not from the `INITIAL` state and to confirm that Provider admins can accept/reject moderated sanctions.

Also cleaned up some of the existing reviews tests and added new tests to confirm that RegistrationActions are written during the moderation flow as a target of opportunity.

## Ticket
https://openscience.atlassian.net/browse/ENG-2378

### CODE REVIEW

**Comment by jwalz** [comment]
With the new tests for RegistrationActions in `test_review_flows`, I feel like the tests for writing actions in `test_registrations` can be removed. Anybody else have thoughts?

File: osf/models/provider.py [hunk 1]
```diff
@@ -234,6 +234,15 @@ def validate_schema(self, schema):
         if not self.schemas.filter(id=schema.id).exists():
             raise ValidationError('Invalid schema for provider.')
 
+    def user_is_moderator(self, user):
+        if not user:
+            return False
+
+        # generator so that we only check both groups if necessary
+        moderator_groups = (self.get_group(group_name) for group_name in ('moderator', 'admin'))
+        return any(group.user_set.filter(id=user.id).exists() for group in moderator_groups)
```
Comment (brianjgeiger) [inline osf/models/provider.py:243]: Drive-by: Could you see if something like `get_perms(user, provider)` would work here? I think it would be more accurate and probably easier. You can import get_perms from guardian.shortcuts.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Quick drive-by.

**Review by UdayVarkhedkar — APPROVED** [review]
Looks good to me! As far as I can tell, the tests you've written in test_review_flows.py makes the writing action ones in osf_tests/test_registrations.py redundant so they can be removed.

File: tests/test_registrations/test_views.py [hunk 2]
```diff
@@ -490,3 +500,102 @@ def test_check_draft_state_registered_and_deleted_and_approved(self):
...
+        self, app, embargoed_registration, moderator, registration_subpath):
+        # Moderators should not have non-standard access to a registration
+        # before it is submitted for moderation by its authors
+        assert embargoed_registration.moderation_state == 'initial'
+
+        resp = app.get(registration_subpath, auth=moderator.auth, expect_errors=True)
+        assert resp.status_code == 403
+
+    @pytest.mark.enable_quickfiles_creation
+    def test_modeartor_can_view_subpath_of_submitted_registration(
```
[the remark this answers is not in the document]
Comment (jwalz) [reply tests/test_registrations/test_views.py]: thanks for the catch

**Comment by jwalz** [comment]
> Looks good to me! As far as I can tell, the tests you've written in test_review_flows.py makes the writing action ones in osf_tests/test_registrations.py redundant so they can be removed.

Yep, removed them

**Review by brianjgeiger — APPROVED** [review]
